### ISSUE TITLE
[Bug] Error in Balance invoices (7.0.x)

### ISSUE THREAD

**Issue by minigitmax** [Dolibarr/dolibarr#8452]
# Bug
Invoices generated from orders with negative records (as for discounts or previous deposit already billed) starting from version 7.0.0 automatically undergo the conversion of negative lines in "discount" (?!). This conversion happens without any warning and strongly condition the invoice issue and payment process, since a global discount is created for the involved third party and the invoice and payment status are consequently altered.
A similar behavior happens when a payment of a deposit invoice is entered. Also in this case, the payment generates a "discount" and the invoice status changes in "converted to discount" instead of "payed". The same happens for Credit Notes issued to cancel previous invoices (i.e. for cancelled orders).

This behavior results in a very confusing and hazardous process. 
The idea of automatizing these kind of invoices/payment processes would be good, but the current implementation looks confused and buggy.

Then there is a **serious bug in the printed output of these kind of invoices (by crabe model)** with auto-generated discounts from negative lines. The negative lines (after discount auto-generation) are not correctly computed as "Net in currency" and so not considered in the Total (net of tax) amount. These final invoices with negative lines show wrong "net of tax" total and correct "inc. tax" total.
This bug has been already reported in the issue #8247.

## Environment
- **Version**: 7.0.1
- **OS**: Linux 3.10.102 #15254 SMP Fri Jan 26 06:40:50 CST 2018 x86_64
- **Web server**: Apache/2.2.34 (Unix) mod_fastcgi/mod_fastcgi-SNAP-0910052141
- **PHP**: 5.6.34
- **Database**: MariaDB 5.5.57

**Comment by eldy** [comment]
What do you mean with "Invoices generated from orders with negative records" ?
Do you mean you make an order with negative lines ? And then you generate an invoice from order ?

**Comment by eldy** [comment]
About comment :
"A similar behavior happens when a payment of a deposit invoice is entered. Also in this case, the payment generates a "discount" and the invoice status changes in "converted to discount" instead of "payed". The same happens for Credit Notes issued to cancel previous invoices (i.e. for cancelled orders)."

This is what we want. When a deposit invoice (Down payment) is payed completely, it is converted into the list of customer "discount"  (here the term "discounts and credits" would be better). It means, it is payed and it is ready to be consumed into another invoice. The short status is "Payed" and long status label is "Payed (ready for consumption in final invoice)".  But you are free to consume it or not, it does not change nothing.
So i don't see what's wrong. Can you provide example with amount so we can reproduce trouble ?

**Comment by minigitmax** [comment]
Hi Eldy.

Regarding your first question: 

> What do you mean with "Invoices generated from orders with negative records" ?
Do you mean you make an order with negative lines ? And then you generate an invoice from order ?

Exactly. Invoices generated from order with negative lines, such as discounts or amounts already paid (deposit, down payments) entered manually.
More details in the next reply to your second comment.

**Comment by minigitmax** [comment]
About your second comment:

> This is want we want. When a deposit invoice (Down payment) is payed completely, it is converted into the list of customer "discount" (here the term "discounts and credits" would be better). It means, it is payed and it is ready to be consumed into another invoice. The short status is "Payed" and long status label is "Payed (ready for consumption in final invoice)". But you are free to consume it or not, it does not change nothing. So i don't see what's wrong. 

The intention is clear and I agree with that. But the realization seems not fully performing and valid for all the possible cases (ok for down payments, not good for credit notes or standard invoices with negative lines). Besides the adopted status names seem not always correct (especially in the italian translation).

**Case of Down payment**
The process is ok, but there is a severe bug in the "Total (net in currency)" field, so the final invoice is wrong. See the next screenshots. Besides, descriptions and names are a bit confusing: 
1) the payment of the deposit invoice generates firstly a credit (correct) that is called _Absolute discount_ in the customer card ("This Deposit invoice has been converted into Discount")
2) the invoice status becomes _Paid_ in English (ok) but _Convertita in sconto_ (italian translation of _Converted into discount_) in Italian
3) the order's _Discounts_ become "This customer has no relative discount by default. This customer still has _credit notes_ for 2,440.00 Euros". Credit Notes aren't discounts nor deposits.
4) again, the draft of the next balance invoice shows "This customer still has _credit notes_ for 2,440 €". If applied, the new invoice status will be _paid_ once paid (ok, also in italian).

![image](https://user-images.githubusercontent.com/26568957/38734581-70378556-3f26-11e8-80cb-e9fdf9246b18.png)
![image](https://user-images.githubusercontent.com/26568957/38734652-af4af2b4-3f26-11e8-8aeb-496acf434ad3.png)

**Case of Credit Notes of Already Paid Invoices**
The process is ok, except again for the terms Discount/Credit Notes used with the same meaning

**Case of Credit Notes of Not Already Paid Invoices**
This kind of credit note is not correctly managed since the available options are the same as for the Credit Notes of Already Paid Invoices, so it's only possible to "Issue a refund" (?!) or to "Convert into discount" (?!): both these options are not suitable for not paid invoices. The available status of the original invoice remains _Not Paid_ that is true in the strict sense but not useful for accounting (the invoice remains unpaid...); the Credit Note status remains _Not paid_ as well and _Non rimborsata_ (the italian translation for _Not refunded_) in italian language.

**Case of Invoices generated from Orders with Negative Lines**
Starting from orders with negative lines, such as orders with special discounts exposed in the order/invoice (ref. the screenshot below), firstly there's the fact that it's not possible to directly add negative lines (but only modify at a second step the quantity or the price of positive values), then there are troubles again in the invoice generation if the negative lines derives from a negative price (discount).

![image](https://user-images.githubusercontent.com/26568957/38738282-06b2e682-3f32-11e8-8408-e5f4e7f75324.png)
![image](https://user-images.githubusercontent.com/26568957/38738350-45ad65a6-3f32-11e8-8a01-3e5c7bcffe48.png)

The negative lines is automatically converted into discount and moved to the first position. When validated, the invoice suffers the "Net in currency" bug above described resulting in a wrong invoice.
An _Already consumed discount_ is automatically generated as well (this is acceptable...).
As per completeness, I must say that if the negative lines are entered by modifying the quantity (not the price) the discount is not automatically generated and the resulting invoice is correct.

**Comment by minigitmax** [comment]
The **bug** and other mentioned issues are still there after updating to 7.0.2

**Comment by eldy** [comment]
Let's discard the translation problem. With v8, a lot of term were changed into english so translation for it has been redone. It may have fixed the trouble.

For the other trouble, can you tell me if you have enable the module "multi currency" ? If yes, can you try to regenerate the invoice after disabling it ?

**Comment by minigitmax** [comment]
Hello eldy,

Yes. Just checked after disabling the "multi currency" module before regenerating the invoice. The new invoice is formally correct. How can we do?

The translation issue may have been solved. I've made only a quick check, and it seems that in English the things are more properly named now. In the standard italian translation there's a trouble with the translation of "Deposit", "Deposit Invoice" that is translated to "Accredito", "Nota d'Accredito". In italian, "Nota d'Accredito", or more frequently "Nota di Credito", is the Credit Note (for reimbursement or deletion of a previous invoice), whereas "Deposit Invoice" (or better "Down Payment Invoice") should be translated as "Fattura d'Acconto". For now I've changed the .lang file, waiting for a fixing.
